### DISCUSSION TITLE
How to use a different `album_path` for compilations?

### DISCUSSION THREAD

**Discussion by ali-ramadhan** [MoeMusic/Moe#231] [Q&A]
I am trying to place compilations (and eventually soundtracks) in separate directories which in beets is achieved by conditioning the album path on the album type. I don't think this is currently possible with Moe but I'm wondering what features are needed to be able to do this.

Not sure how specifying different album paths under certain criteria should work. Can a plugin be used to do this? The beets approach of specifying conditional album paths in the configuration file seems fine to me but it might not necessarily be the best or most flexible.

The other missing piece is having the album type be a `MetaAlbum` and `Album` property which can be grabbed from Musicbrainz. This should be pretty straightforward to implement I think? Happy to give it a shot if so. It should then be easy to implement the functionality of the beets `albumtypes` plugin using Moe plugins.

**Comment by jtpavlock** [comment]
Could you give some examples of your desired end-result paths?

**Reply by ali-ramadhan** [reply]
Was mostly thinking about compilations for now but here are some examples of what I was thinking of doing with multiple album paths.

Album with one artist (default path):
```
Artists/M/My Chemical Romance/My Chemical Romance - The Black Parade (2006)/
```

If the album type is compilation then

```
Compilations/R/Rough Guide to the Music of the Sahara, The (2005)/
```

If the album type is soundtrack then

```
Soundtracks/P/Pulp Fiction: Music from the Motion Picture (1994)/
```

If classical is in genres then

```
Classical/D/Dvořák, Antonín (1841-1904)/Symphonies Nos. 7, 8 & 9 (1994) performed by London Philharmonic Orchestra under Carlo Maria Giulini/
```

**Reply by jtpavlock** [reply]
Okay, since your path structure is roughly the same, could you just create a function that changes the first part of your album path? So something like:

`artist_switcher(album)/...`

That returns either the album artist, compilation, etc?

You're right though, we'd need a new album type field.

Edit: I see the extra sub path you want for artists now. You could still do this with functions, but it'd get messy. I think there should/can be a better way. Let me think about how to best handle this.

**Reply by ali-ramadhan** [reply]
Ah an `artist_switcher` plugin should work but yeah it could get messy, especially if you want to use it to handle classical music in a different manner.

A potential idea building on https://github.com/MoeMusic/Moe/discussions/229#discussioncomment-4052453: If there were hooks that could modify the album path both before and after splitting (or something like `override_album_path_template`) then all the conditional switching can happen in the plugin.

**Reply by jtpavlock** [reply]
Yeah I think creating an `override_album_path_template` hook is going to be the best bet. The problem with the before and after splitting evaluation is that it would be much more difficult to implement.

**Reply by jtpavlock** [reply]
If you're interested in implementing the new hook, it shouldn't be too hard and I can layout the steps if you like.

**Reply by ali-ramadhan** [reply]
Happy to give it a shot! Any tips on how to start would be great.

**Reply by jtpavlock** [reply]
Alright! I think the following steps should cover most of it, but don't hesitate to ask if you have any questions:

1. Create a new `hookspec` in the `move_core.Hooks` class (follow the `create_path_template_func` one as an example). I think the name should actually be `override_album_path_config` to be more explicit on what is being overriden.
    a. Try to include an example of how to use the hook in the docstring of the hookspec. 
2. In `test_move_core.py` add a hook implementation to the `MovePlugin` class and then add a new test class for testing the hook itself. Testing should probably just call `fmt_item_path(album) and test the path is changed per the hook implementation you added.
3. In `_fmt_album_path()` you should execute the hook and see if there are any results before defaulting to the configuration setting. Running the hook is accomplished by calling `config.CONFIG.pm.hook.override_album_path_config()`.
4. There should be some allusion to the new hook in the `move.rst` docs. Maybe we need some advanced configuration section that talks about creating a local plugin to implement one or both of `create_path_template_func` and `override_album_path_config` hooks linking to both. Not sure what the best approach is here, but whatever you think would be helpful for someone new going through the docs and trying to extend the functionality of their path templates.

(steps 2 and 3 can be swapped based on your preference of TDD)

**Reply by ali-ramadhan** [reply]
Sorry I haven't had a chance to give this a try yet, been busy with work and other obligations. Hoping to revisit when I have more free time some time in the next few weeks!

**Reply by jtpavlock** [reply]
No worries! Thanks for checking back in, I certainly understand how open source hobbies need to take a back seat to real life at times.

**Comment by ali-ramadhan** [comment]
Hey @jtpavlock! Taking another stab at music tagging so I'm revisiting this discussion after almost 2 years lol.

I'm thinking of opening a PR implementing your suggestion for an `override_album_path_config` hook (from your comment here: https://github.com/orgs/MoeMusic/discussions/231#discussioncomment-4069614).

Just feeling a bit rusty. Could you just remind/explain how this hook differs from `create_path_template_func`? Or do you think a different solution would be better now?

I know you explained the idea a bit more in https://github.com/orgs/MoeMusic/discussions/229#discussioncomment-4052613. Is the main idea that `override_album_path_config` would use a different `_eval_path_template` function that doesn't sanitize first (or at all)?

If it doesn't sanitize that might be good because sanitation could be manually invoked on e.g. artist names in case they have `/`?

**Reply by jtpavlock** [reply]
Welcome back @ali-ramadhan! 

I honestly had to refresh myself as well on the ideas we talked about. It seems like you have it correct though i.e. the `create_path_template_func` provides access to functions that evaluate prior to path sanitization, so there's no ability to specify custom sub-directories within them as any '/' would get sanitized out, while the proposed `override_album_path_config` would allow the user to provide different path templates entirely based on custom logic.

So I think it should work something like this:

In your plugin:
```python
@moe.hookimpl
def override_album_path_config(album):
   if album is soundtrack:
      return "Soundtracks/{album.artist}/{album.title}"  <--- path template format string
   else:
      return "Albums/{album.artist}/{album.title}"  <--- different path template format string
```

Then in moe, in `move_core.fmt_item_path`, call the hook, and change the album `item_path` to call `_eval_path_template` on either the configuration value, or the returned path string from the `override_album_path_config` if it exists.

Let me know if you think this would work or if you have other ideas.

**Comment by jtpavlock** [comment]
Resolved with b062fc55d5a1a90bc4451504330e4edefad0e3a7
